File: JavaScript/110.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
var isBalanced = function (root) {
  return dfs(root) !== -1;
};

const dfs = (node) => {
  if (!node) {
    return 0;
  }

  let leftHeight = dfs(node.left);
  if (leftHeight === -1) return -1;

  let rightHeight = dfs(node.right);
  if (rightHeight === -1) return -1;

  if (Math.abs(leftHeight - rightHeight) > 1) {
    return -1;
  }

  return 1 + Math.max(leftHeight, rightHeight);
};

File: Rust/110.rs
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
//
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::cell::RefCell;
use std::cmp;
use std::rc::Rc;
impl Solution {
    pub fn is_balanced(root: Option<Rc<RefCell<TreeNode>>>) -> bool {
        Self::dfs(&root) != -1
    }

    fn dfs(node: &Option<Rc<RefCell<TreeNode>>>) -> i32{
        let node_ref = match node {
            Some(n) => n.borrow(),
            None => return 0
        };

        let left_height = Self::dfs(&node_ref.left);
        if left_height == -1 {
            return -1;
        }
        let right_height = Self::dfs(&node_ref.right);
        if (right_height == -1) {
            return -1;
        }

        if (left_height - right_height).abs() > 1 {
            return -1;
        }

        1 + cmp::max(left_height, right_height)
    }
}
